ack_timer.go

```diff
@@ -41,34 +52,35 @@ func (t *ackTimer) start() bool {
 	t.mutex.Lock()
 	defer t.mutex.Unlock()
 
-	// this timer is already closed
-	if t.closed {
+	// this timer is already closed or already running
+	if t.state != ackTimerStopped {
 		return false
 	}
-
-	// this is a noop if the timer is already running
-	if t.stopFunc != nil {
-		return false
-	}
-
-	cancelCh := make(chan struct{})
+	t.state = ackTimerStarted
```

Comment:
Is there a chance of race here? Is there a need to drain timer channel or t.cancel?

Thinking aloud
1. `stop()` called: that will write to `t.cancel` and set state to `stopped`
2. Assume, the go routine below has not drained `t.cancel` yet.
3. `start()` happens: this will try to restart the timer as state is `stopped`
4. This will launch another go routine. Will this clash with go routine in Step 2?

Reply:
good catch. invocations of the callback are interchangeable and for this to happen the time between calls has to be practically zero so we can eat the extra cancel and the earlier goroutine will serve as the timeout for the latest call to start.